### ISSUE TITLE
Automatic r_to_py conversion fails on {arrow} objects

### ISSUE THREAD

**Issue by nealrichardson** [rstudio/reticulate#748]
I'm working on a [vignette](https://github.com/apache/arrow/blob/ff1f58c9a38d4f5d12963bbda82ef37262f30231/r/vignettes/python.Rmd) for the `arrow` package to document its [new `reticulate` support](https://github.com/apache/arrow/blob/master/r/R/py-to-r.R) for passing data to/from Python with `pyarrow`. In the process, I observed some unexpected behavior.

1. While the automatic `r_to_py` conversion works for regular R objects, it seems to fail for Arrow objects (R6 classes that hold an externalptr). Assuming `arrow` and `pyarrow` as described in that vignette (you need the nightly version of both packages as this is not yet released), this first part works as expected:

```r
pa <- import("pyarrow")
a <- pa$array(c(1, 2, 3))
a

## Array
## <double>
## [
##   1,
##   2,
##   3
## ]
```

I created an Arrow Array using `pyarrow`, passing an R `numeric` vector that gets automatically translated to Python, and because I did `import()` with the default `convert = TRUE`, the result is an R Arrow Array, as if I had made it directly with the R package. All good here.

But if the inputs to the Python function are an Arrow R6 class object, wrapping a C++ pointer, they don't get automatically converted with `r_to_py`. 

```r
# Creating arrays using the R package
a <- Array$create(c(1, 2, 3))
b <- Array$create(c(6, 7, 8))
# Now call a function in pyarrow
pa$concat_arrays(list(a, b))

## Error in py_call_impl(callable, dots$args, dots$keywords) : 
##   Unable to access object (object is from previous session and is now invalid)
```

However, if I explicitly wrap the function input in `r_to_py`, it works:

```r
pa$concat_arrays(r_to_py(list(a, b)))

## Array
## <double>
## [
##   1,
##   2,
##   3,
##   6,
##   7,
##   8
## ]
```

2. If I import with `convert = FALSE`, which I understood to only affect the `py_to_r` behavior, now the automatic `r_to_py` works.

```r
# (in a fresh session)
pa <- import("pyarrow", convert = FALSE)
a <- Array$create(c(1, 2, 3))
b <- Array$create(c(6, 7, 8))
pa$concat_arrays(list(a, b))

## [
##   1,
##   2,
##   3,
##   6,
##   7,
##   8
## ]
```

I'm not very familiar with the internals of `reticulate`, but if I had to guess, I wonder if the code that's inspecting the function inputs sees the Arrow externalptrs, sees that they don't map onto any Python objects from this R session, and assumes that they're stale Python refs.

**Comment by skeydan** [comment]
Just a WIP comment... yeah for me too here `convert`, as set on the R object, makes all the difference...

```
library(reticulate)
library(arrow)
pa <- import("pyarrow")

a <- Array$create(c(1, 2, 3))

# crashes
x <- pa$concat_arrays(list(a))

# works
x <- pa$concat_arrays(list(r_to_py(a, T)))
```
To be continued....

**Comment by skeydan** [comment]
Sorry not getting very far, apart from finding (unsurprisingly) that when this crashes, here

https://github.com/rstudio/reticulate/blob/330ea87165c1812199c1bb9198e7f43d5f94bd9f/src/reticulate_types.h#L29

`pyobj` is an `R_NilValue` where it shouldn't be...

Just one weird speculation... Could it be a problem that in the calling R script, `pyarrow` is imported with `convert = TRUE` (the default), while in `r_to_py.Array` it's used with `FALSE`,

```
function (x, convert = FALSE) 
{
  schema_ptr <- allocate_arrow_schema()
  array_ptr <- allocate_arrow_array()
  on.exit({
    delete_arrow_schema(schema_ptr)
    delete_arrow_array(array_ptr)
  })
  pa <- reticulate::import("pyarrow", convert = convert)
  ExportArray(x, array_ptr, schema_ptr)
  pa$Array$`_import_from_c`(array_ptr, schema_ptr)
}
```

thus creating some inconsistency? Just speculating (and the error goes away if we import with `convert = FALSE` in the calling script)

**Comment by nealrichardson** [comment]
You surely know better than I, but I don't see the connection. IIUC, the issue is that the (automatic) call to `r_to_py` isn't dispatching to `r_to_py.Array` at all, not that it's being called with inconsistent options. In any case we can certainly experiment with editing the arrow r_to_py/py_to_r methods if you think it could help.

My theory (just from trying to read the source, not from stepping through a debugger, so maybe you've ruled this out) is that somewhere higher up in the C++ code, it is assuming the arrow externalptr must be a pyobject, so it's not calling back to the `r_to_py.Array` R S3 method like it does for other R objects, and it's instead going through the pyobject path.

**Comment by kevinushey** [comment]
FWIW I'm just seeing:

```
> pa$concat_arrays(list(a, b))
Error in py_get_attr_impl(x, name, silent) :
  AttributeError: module 'pyarrow' has no attribute 'concat_arrays'
```

**Comment by nealrichardson** [comment]
@kevinushey I'm guessing you have an old version of pyarrow somehow then. Did you `arrow::install_pyarrow()` to get it? What is the `__version__` you have?

**Comment by skeydan** [comment]
> IIUC, the issue is that the (automatic) call to` r_to_py` isn't dispatching to `r_to_py.Array` at all, not that it's being called with inconsistent options. 

No, I see that it is actually getting called (which is how I came to see its definition :-)). But still, I'm happing @kevinushey is jumping in (((I installed the nightly builds as described in some vignette)))

**Comment by kevinushey** [comment]
Okay, this seems to be an MRE:

```
library(reticulate)
library(arrow)

pa <- import("pyarrow")

# Creating arrays using the R package
a <- Array$create(c(1, 2, 3))
b <- Array$create(c(6, 7, 8))
r_to_py(list(a, b), convert = TRUE)
```

It looks like `r_to_py(a, convert = TRUE)` doesn't actually return a Python object:

```
> class(r_to_py(a, convert = TRUE))
[1] "Array"       "ArrowObject" "R6" 
```

I think this implies that the `convert` flag is being used incorrectly here:

https://github.com/apache/arrow/blob/1f175e12a101960fd5b293a2dc366855eaaad141/r/R/python.R#L38

That should likely be `convert = FALSE`, regardless of the value of `convert` passed in, since you always want to get a Python object back.

The `convert` flag here isn't really well documented. What `r_to_py(..., convert = TRUE)` really controls is whether wrapped Python functions should auto-convert their return values. For example:

```
> library(reticulate)
> sys1 <- import("sys", convert = FALSE)
> sys2 <- import("sys", convert = TRUE)
> 
> class(sys1)
[1] "python.builtin.module" "python.builtin.object"
> class(sys2)
[1] "python.builtin.module" "python.builtin.object"
> 
> class(sys1$path)
[1] "python.builtin.list"   "python.builtin.object"
> class(sys2$path)
[1] "character"
```

In other words, if you're returning a Python object that might contain Python functions, the convert flag is used to control whether those functions should auto-convert their results to R, or return Python objects.

**Comment by nealrichardson** [comment]
Ah, I see (I think): `pa$concat_arrays()` calls `r_to_py()` on its inputs with `convert = TRUE` because `pa` is imported with `convert = TRUE`. But `r_to_py.Array()` should ignore that and always use `FALSE` because the whole point of the method is to return a Python object. 

One thing that doesn't seem quite right though is that now I'm returning a Python object with `convert = FALSE`, so any Python methods it has will return Python objects, not R objects like everything else I get from the module I imported with `convert = TRUE`. So should I (can I) assign the "actual" convert value back, like this?

```diff
diff --git a/r/R/python.R b/r/R/python.R
index 4d84a4083..e6782689e 100644
--- a/r/R/python.R
+++ b/r/R/python.R
@@ -35,9 +35,11 @@ r_to_py.Array <- function(x, convert = FALSE) {
     delete_arrow_array(array_ptr)
   })
 
-  pa <- reticulate::import("pyarrow", convert = convert)
+  pa <- reticulate::import("pyarrow", convert = FALSE)
   ExportArray(x, array_ptr, schema_ptr)
-  pa$Array$`_import_from_c`(array_ptr, schema_ptr)
+  out <- pa$Array$`_import_from_c`(array_ptr, schema_ptr)
+  out$convert <- convert
+  out
 }
 ```

**Comment by nealrichardson** [comment]
Ok, thanks for working through that with me. Can I commemorate what we learned here with a PR to improve documentation? Should there be some language added somewhere around https://github.com/rstudio/reticulate/blob/master/vignettes/package.Rmd#L137? Is there a better/additional place?

**Comment by kevinushey** [comment]
That makes sense to me. I'll update the documentation.
